perf(transformer/using): inline `enter_statement`

overlookmotel opened the pull request:
`enter_statement` is a hot path, and most `Statement`s are not `BlockStatement`s or `SwitchStatement`s, so in most cases this method does nothing.

Optimize for the common "nothing to do here" case by moving as much logic as possible out of `enter_statement`, to make that function as small as possible, so it can be inlined. Mark it `#[inline]`.

Hopefully now when a `Statement` is neither of the types this visitor is interested in, finding that out and exiting will just be a couple of quick checks, and won't incur the cost of a function call.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9680?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9680** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9680?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9680?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9678** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9678?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9677** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9677?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9676** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9676?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9675** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9675?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9674** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9674?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9673** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9673?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9672** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9672?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9680).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented:
Come on CodSpeed! Where's your report?

overlookmotel commented:
<img width="775" alt="bench" src="https://github.com/user-attachments/assets/a082bfb3-86a9-4b78-9e0e-55232f37bdc5" />

A small gain. I guess most of the cost of #9310 is iterating over statements, as we suspected.

Boshen commented:
### Merge activity

* **Mar 11, 5:53 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
perf(transformer/using): inline `enter_statement` (#9680)

`enter_statement` is a hot path, and most `Statement`s are not `BlockStatement`s or `SwitchStatement`s, so in most cases this method does nothing.

Optimize for the common "nothing to do here" case by moving as much logic as possible out of `enter_statement`, to make that function as small as possible, so it can be inlined. Mark it `#[inline]`.

Hopefully now when a `Statement` is neither of the types this visitor is interested in, finding that out and exiting will just be a couple of quick checks, and won't incur the cost of a function call.
